wallet/src/de/schildbach/wallet/ExploreSyncWorker.kt

```diff
@@ -75,91 +71,78 @@ class ExploreSyncWorker constructor(val appContext: Context, workerParams: Worke
-                dao.deleteAll(source)
-                syncTable(tableName, valueType, dao)
-                preferences.edit().putLong(prefsLastSyncKey, lastDataUpdate).apply()
-                log.info("Sync $tableName finished")
+            tableSyncWatch.reset()
+            tableSyncWatch.start()
+
+            val merchantDataSize = exploreRepository.getMerchantDataSize()
+            if (merchantDataSizeDB == 0) {
+                merchantDao.deleteAll()
             } else {
-                log.info("Data $tableName timestamp $lastSync, nothing to sync (${Date(lastSync)})")
+                log.info("MERCHANT data partially synced ($merchantDataSizeDB of $merchantDataSize), continuing")
             }
-            true
-        } catch (ex: Exception) {
-            log.error("Error while syncing ${tableName}: ${ex.message}")
-            analytics.logError(ex, tableName)
-            false
-        }
-    }
+            val merchantCache = mutableListOf<Merchant>()
+            var counter = 0
+            exploreRepository.getMerchantData(merchantDataSizeDB).collect {
+                counter++
+                merchantCache.add(it)
+                if (merchantCache.size == PAGE_SIZE) {
+                    merchantDao.save(merchantCache)
+                    merchantCache.clear()
+                    log.info("MERCHANT $counter of ${merchantDataSize - merchantDataSizeDB} sync took $tableSyncWatch")
+                }
+            }
+            if (merchantCache.size > 0) {
+                merchantDao.save(merchantCache)
+                merchantCache.clear()
+            }
+            tableSyncWatch.stop()
+            log.info("MERCHANT sync took $tableSyncWatch ($counter records)")
 
-    private suspend fun <T: SearchResult> syncTable(
-        tableName: String,
-        valueType: Class<T>,
-        dao: BaseDao<T>
-    ) {
-        val dataSize = exploreRepository.getDataSize(tableName)
-        val totalPages = ceil(dataSize.toDouble() / PAGE_SIZE).toInt()
-        log.info("$tableName $dataSize records in $totalPages chunks")
-        val tableSyncWatch = Stopwatch.createStarted()
-        var synced = 0
-
-        for (page in 0 until totalPages) {
-            val startAt = page * PAGE_SIZE
-            val endAt = startAt + PAGE_SIZE
-            log.info("$tableName chunk ${page + 1} of $totalPages ($startAt to $endAt)")
-            val data = exploreRepository.get(tableName, startAt, endAt, valueType)
-            synced += data.size
-            dao.save(data)
-        }
+            preferences.edit()
+                .putLong(PREFS_LAST_SYNC_KEY, Calendar.getInstance().timeInMillis)
+                .apply()
 
-        if (synced != dataSize) {
-            throw FirebaseException("sync is broken, expected $dataSize records, got $synced")
+            log.info("Sync Explore Dash finished")
+
+            return@withContext Result.success()
+
+        } catch (ex: Exception) {
+            analytics.logError(ex)
```

Comment:
It's better if we log some details here, otherwise, it'll be hard to say from crashlytics whether the worker is interrupted just once or several times. I suggest logging something like "syncing from ${merchantDataSizeDB}"

Reply:
Good point, fixed